go/runtime/txpool/txpool.go

```diff
@@ -231,38 +234,46 @@ func (t *txPool) SubmitTxNoWait(ctx context.Context, tx []byte, meta *Transactio
 }
 
 func (t *txPool) submitTx(ctx context.Context, rawTx []byte, meta *TransactionMeta, notifyCh chan *protocol.CheckTxResult) error {
-	tx := newTransaction(rawTx, txStatusPendingCheck)
+	// todo: in the new design, we'll be submitting to the check queue first, which won't have "first seen time"
+	// metadata. seems like it could still work out, as checking is single threaded, so we can still preserve the
+	// original order when we finish checking and offer to the main queue
 
+	// todo: can we go without this?
 	// Skip recently seen transactions.
-	if _, seen := t.seenCache.Peek(tx.hash); seen {
-		t.logger.Debug("ignoring already seen transaction", "tx_hash", tx.hash)
+	// todo: hash used to be cached
+	h := hash.NewFromBytes(rawTx)
+	if _, seen := t.seenCache.Peek(h); seen {
+		t.logger.Debug("ignoring already seen transaction", "tx_hash", h)
 		return fmt.Errorf("duplicate transaction")
 	}
 
 	// Queue transaction for checks.
 	pct := &PendingCheckTransaction{
-		Transaction: tx,
-		notifyCh:    notifyCh,
-	}
-	if meta.Local {
-		pct.flags |= txCheckLocal
+		tx:       rawTx,
+		notifyCh: notifyCh,
 	}
 	if meta.Discard {
-		pct.flags |= txCheckDiscard
+		pct.dstQueue = nil
+	} else if meta.Local {
+		pct.dstQueue = t.localQueue
+	} else {
+		pct.dstQueue = t.mainQueue
 	}
 
 	return t.addToCheckQueue(pct)
 }
 
 func (t *txPool) addToCheckQueue(pct *PendingCheckTransaction) error {
+	// todo: hash used to be cached
+	h := hash.NewFromBytes(pct.tx)
 	t.logger.Debug("queuing transaction for check",
 		"tx", pct.tx,
-		"tx_hash", pct.hash,
-		"recheck", pct.isRecheck(),
+		"tx_hash", h,
+		"recheck", "dunno", // todo: can we live without this?
```

Comment:
It is useful if we know at any point whether a given transaction has already been checked by the runtime. Not just for logging, but this can also be used as an optimization (later), for example it can hint the runtime that it does not need to check signatures on re-checks as they have already been verified in the initial check.

Why is this a problem?

Reply:
let's preserve it for use in such future optimizations then. good that you bring up the idea, since I didn't see that in examining existing code

Reply:
pct.flags is back, with a bit for this recheck flag